internal/core/operations/healthcontroller/health_controller_executor.go

```diff
@@ -0,0 +1,250 @@
+	}
+	instances, err = ex.instanceStorage.GetAllInstances(ctx, op.SchedulerName)
+	if err != nil {
+		logger.Error("error fetching instances")
+		return gameRoomIDs, instances, scheduler, err
+	}
+
+	scheduler, err = ex.schedulerStorage.GetScheduler(ctx, op.SchedulerName)
+	if err != nil {
+		return gameRoomIDs, instances, scheduler, err
+	}
+	return gameRoomIDs, instances, scheduler, err
+}
+
+func (ex *SchedulerHealthControllerExecutor) checkNonexistentGameRoomsIDs(gameRoomIDs []string, gameRoomInstances []*game_room.Instance) []string {
+	var nonexistentGameRoomsIDs []string
+	for _, gameRoomID := range gameRoomIDs {
+		found := false
+		for _, instance := range gameRoomInstances {
+			if instance.ID == gameRoomID {
+				found = true
+				break
+			}
+		}
+		if !found {
+			nonexistentGameRoomsIDs = append(nonexistentGameRoomsIDs, gameRoomID)
+		}
+	}
+	return nonexistentGameRoomsIDs
+}
+
+func (ex *SchedulerHealthControllerExecutor) tryEnsureCorrectRoomsOnStorage(ctx context.Context, op *operation.Operation, logger *zap.Logger, nonexistentGameRoomIDs []string) {
+	for _, gameRoomID := range nonexistentGameRoomIDs {
+		err := ex.roomStorage.DeleteRoom(ctx, op.SchedulerName, gameRoomID)
+		if err != nil {
+			msg := fmt.Sprintf("could not delete nonexistent room %s from storage", gameRoomID)
+			logger.Warn(msg, zap.Error(err))
+			continue
+		}
+		logger.Sugar().Infof("remove nonexistent room on storage: %s", gameRoomID)
+	}
+}
+
+func (ex *SchedulerHealthControllerExecutor) ensureDesiredAmountOfInstances(ctx context.Context, op *operation.Operation, logger *zap.Logger, actualAmount, desiredAmount int) error {
+	var msgToAppend string
+
+	if actualAmount > desiredAmount {
+		removeAmount := actualAmount - desiredAmount
+		removeOperation, err := ex.operationManager.CreatePriorityOperation(ctx, op.SchedulerName, &remove_rooms.RemoveRoomsDefinition{
+			Amount: removeAmount,
+		})
+		if err != nil {
+			return err
+		}
+		msgToAppend = fmt.Sprintf("created operation (id: %s) to remove %v rooms.", removeOperation.ID, removeAmount)
+	} else {
+		addAmount := desiredAmount - actualAmount
+		addOperation, err := ex.operationManager.CreatePriorityOperation(ctx, op.SchedulerName, &add_rooms.AddRoomsDefinition{
+			Amount: int32(addAmount),
+		})
+		if err != nil {
+			return err
+		}
+		msgToAppend = fmt.Sprintf("created operation (id: %s) to add %v rooms.", addOperation.ID, addAmount)
+	}
+
+	logger.Info(msgToAppend)
+	ex.operationManager.AppendOperationEventToExecutionHistory(ctx, op, msgToAppend)
+	return nil
+}
+
+func (ex *SchedulerHealthControllerExecutor) findAvailableAndExpiredRooms(ctx context.Context, op *operation.Operation, gameRoomsIDs []string) (availableRoomsIDs, expiredRoomsIDs []string) {
+	for _, gameRoomID := range gameRoomsIDs {
+		room, err := ex.roomStorage.GetRoom(ctx, op.SchedulerName, gameRoomID)
+		if err != nil {
+			continue
+		}
+
+		if ex.isRoomStatusError(room) {
+			continue
+		}
+
+		if ex.isRoomExpired(room) {
```

Comment:
I think that here there is a possibility of the status of the room being `terminating`, so it should not be considered an expired one. WDYT?

Reply:
To fix this one I think that you should first ask whether the room is `terminating`, so not asks if it is expired.